tls/s2n_server_hello_retry.c

```diff
@@ -114,15 +116,22 @@ int s2n_server_hello_retry_recv(struct s2n_connection *conn)
         for (size_t i = 0; i < kem_pref->tls13_kem_group_count; i++) {
             if (kem_pref->tls13_kem_groups[i] == kem_group) {
                 match_found = true;
-                ENSURE_POSIX(conn->secure.client_kem_group_params[i].kem_params.private_key.data == NULL,
-                        S2N_ERR_INVALID_HELLO_RETRY);
-                ENSURE_POSIX(conn->secure.client_kem_group_params[i].ecc_params.evp_pkey == NULL,
-                        S2N_ERR_INVALID_HELLO_RETRY);
+                new_key_share_requested = (conn->secure.client_kem_group_params[i].kem_params.private_key.size == 0)
+                        && (conn->secure.client_kem_group_params[i].ecc_params.evp_pkey == NULL);
+                break;
             }
         }
+        ENSURE_POSIX(match_found, S2N_ERR_INVALID_HELLO_RETRY);
     }
 
-    ENSURE_POSIX(match_found, S2N_ERR_INVALID_HELLO_RETRY);
+    /*
+     *= https://tools.ietf.org/rfc/rfc8446#section-4.1.4
+     *# Clients MUST abort the handshake with an
+     *# "illegal_parameter" alert if the HelloRetryRequest would not result
+     *# in any change in the ClientHello.
+     */
+    ENSURE_POSIX(conn->early_data_state == S2N_EARLY_DATA_REJECTED || new_key_share_requested,
```

Comment:
This ensure ~ feels ~ like it could use some parenthesis to make operator precedence more clear.